src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  const stateClone = { ...state };
+
+  for (const action of actions) {
+    const { extraData, type, keysToRemove } = action;
+
+    switch (type) {
+      case 'clear':
+        for (const key in stateClone) {
+          delete stateClone[key];
+        }
+        break;
+
+      case 'addProperties':
+        Object.assign(stateClone, extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of keysToRemove) {
+          delete stateClone[key];
+        }
+        break;
+
+      default:
+        throw new Error(`Wrong type`);
+    }
+
+    states.push({ ...stateClone });
```

Comment:
It is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after the current action is processed. This way, you ensure that the state is only pushed after all actions have been considered for each iteration. You're already doing this correctly, but make sure to keep it consistent if any changes are made to the loop structure.